crates/virtio-queue/src/mock.rs

```diff
@@ -379,62 +407,78 @@ impl<'a, M: GuestMemory> MockSplitQueue<'a, M> {
-    // The `addr` fields of the input descriptors should start at a sufficiently
-    // greater location (i.e. 1MiB, or `0x10_0000`).
-    pub fn add_desc_chain(&self, descs: &[Descriptor], offset: u16) -> Result<(), Error> {
+    pub fn build_desc_chain(&self, descs: &[Descriptor]) -> Result<DescriptorChain<&M>, MockError> {
+        let mut modified_descs: Vec<Descriptor> = Vec::with_capacity(descs.len());
         for (idx, desc) in descs.iter().enumerate() {
-            let i = idx as u16 + offset;
-            let addr = desc.addr().0;
-            let len = desc.len();
             let (flags, next) = if idx == descs.len() - 1 {
                 // Clear the NEXT flag if it was set. The value of the next field of the
                 // Descriptor doesn't matter at this point.
                 (desc.flags() & !VIRTQ_DESC_F_NEXT, 0)
             } else {
                 // Ensure that the next flag is set and that we are referring the following
                 // descriptor. This ignores any value is actually present in `desc.next`.
-                (desc.flags() | VIRTQ_DESC_F_NEXT, i + 1)
+                (desc.flags() | VIRTQ_DESC_F_NEXT, idx as u16 + 1)
             };
-
-            let desc = Descriptor::new(addr, len, flags, next);
-            self.desc_table().store(i, desc)?;
+            modified_descs.push(Descriptor::new(desc.addr().0, desc.len(), flags, next));
         }
+        self.build_multiple_desc_chains(&modified_descs[..])
+    }
 
-        // Update the first available ring position.
+    /// Adds descriptor chains to the memory object of the queue.
+    // `descs` represents a slice of `Descriptor` objects which are used to populate the chains, and
+    // `offset` is the index in the descriptor table where the chains should be added.
+    // The descriptor chain related information is written in memory starting with address 0.
+    // The `addr` fields of the input descriptors should start at a sufficiently
+    // greater location (i.e. 1MiB, or `0x10_0000`).
+    pub fn add_desc_chains(&self, descs: &[Descriptor], offset: u16) -> Result<(), MockError> {
+        let mut new_entries = 0;
         let avail_idx: u16 = self
             .mem
             .read_obj(self.avail_addr().unchecked_add(2))
             .unwrap();
-        self.mem
-            .write_obj(
-                offset,
-                self.avail_addr().unchecked_add(
-                    VIRTQ_AVAIL_RING_HEADER_SIZE + avail_idx as u64 * VIRTQ_AVAIL_ELEMENT_SIZE,
-                ),
-            )
-            .unwrap();
+
+        for (idx, desc) in descs.iter().enumerate() {
+            let i = idx as u16 + offset;
+            self.desc_table().store(i, *desc)?;
+
+            if desc.flags() == VIRTQ_DESC_F_NEXT || idx == 0 {
+                // Update the available ring position.
+                self.mem
+                    .write_obj(
+                        offset + new_entries,
```

Comment:
I think this should be `offset + idx` (or `i`), the entry from the available ring must be the index in the descriptor table of the chain head.

Reply:
Hmm I think this is now `offset + i`, what I meant is offset + idx = i. The explanation is that you need to write in the available ring the index of the descriptor chain head from the descriptor table, and you are adding these descriptors at indexes: offset, offset + 1, offset + 2 ..., where idx = 0, 1, 2 ...

Reply:
Done.